com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/ShaderPassDecal.hlsl

```diff
@@ -0,0 +1,306 @@
+#endif
+
+#if (SHADERPASS == SHADERPASS_DBUFFER_PROJECTOR) || (SHADERPASS == SHADERPASS_FORWARD_EMISSIVE_PROJECTOR) || (SHADERPASS == SHADERPASS_DECAL_SCREEN_SPACE_PROJECTOR) || (SHADERPASS == SHADERPASS_DECAL_GBUFFER_PROJECTOR)
+#define DECAL_PROJECTOR
+#endif
+
+#if (SHADERPASS == SHADERPASS_DBUFFER_MESH) || (SHADERPASS == SHADERPASS_FORWARD_EMISSIVE_MESH) || (SHADERPASS == SHADERPASS_DECAL_SCREEN_SPACE_MESH) || (SHADERPASS == SHADERPASS_DECAL_GBUFFER_MESH)
+#define DECAL_MESH
+#endif
+
+#if (SHADERPASS == SHADERPASS_DBUFFER_PROJECTOR) || (SHADERPASS == SHADERPASS_DBUFFER_MESH)
+#define DECAL_DBUFFER
+#endif
+
+#if (SHADERPASS == SHADERPASS_DECAL_SCREEN_SPACE_PROJECTOR) || (SHADERPASS == SHADERPASS_DECAL_SCREEN_SPACE_MESH)
+#define DECAL_SCREEN_SPACE
+#endif
+
+#if (SHADERPASS == SHADERPASS_DECAL_GBUFFER_PROJECTOR) || (SHADERPASS == SHADERPASS_DECAL_GBUFFER_MESH)
+#define DECAL_GBUFFER
+#endif
+
+#if (SHADERPASS == SHADERPASS_FORWARD_EMISSIVE_PROJECTOR) || (SHADERPASS == SHADERPASS_FORWARD_EMISSIVE_MESH)
+#define DECAL_FORWARD_EMISSIVE
+#endif
+
+#if ((!defined(_MATERIAL_AFFECTS_NORMAL) && defined(_MATERIAL_AFFECTS_ALBEDO)) || (defined(_MATERIAL_AFFECTS_NORMAL) && defined(_MATERIAL_AFFECTS_NORMAL_BLEND))) && (defined(DECAL_SCREEN_SPACE) || defined(DECAL_GBUFFER))
+#define DECAL_RECONSTRUCT_NORMAL
+#elif defined(DECAL_ANGLE_FADE)
+#define DECAL_LOAD_NORMAL
+#endif
+
+#if defined(DECAL_LOAD_NORMAL)
+#include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/DeclareNormalsTexture.hlsl"
+#endif
+
+#if defined(DECAL_PROJECTOR) || defined(DECAL_RECONSTRUCT_NORMAL)
+#include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/DeclareDepthTexture.hlsl"
+#endif
+
+#ifdef DECAL_MESH
+#include "Packages/com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/DecalMeshBiasTypeEnum.cs.hlsl"
+#endif
+#ifdef DECAL_RECONSTRUCT_NORMAL
+#include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/NormalReconstruction.hlsl"
+#endif
+
+void MeshDecalsPositionZBias(inout Varyings input)
+{
+#if UNITY_REVERSED_Z
+    input.positionCS.z -= _DecalMeshDepthBias;
+#else
+    input.positionCS.z += _DecalMeshDepthBias;
+#endif
+}
+
+void InitializeInputData(Varyings input, float3 positionWS, half3 normalWS, half3 viewDirectionWS, inout InputData inputData)
+{
+    inputData.positionWS = positionWS;
+    inputData.normalWS = normalWS;
+    inputData.viewDirectionWS = viewDirectionWS;
+
+#if defined(VARYINGS_NEED_SHADOW_COORD) && defined(REQUIRES_VERTEX_SHADOW_COORD_INTERPOLATOR)
+    inputData.shadowCoord = input.shadowCoord;
+#elif defined(MAIN_LIGHT_CALCULATE_SHADOWS)
+    inputData.shadowCoord = TransformWorldToShadowCoord(positionWS);
+#else
+    inputData.shadowCoord = float4(0, 0, 0, 0);
+#endif
+
+#ifdef VARYINGS_NEED_FOG_AND_VERTEX_LIGHT
+    inputData.fogCoord = half(input.fogFactorAndVertexLight.x);
+    inputData.vertexLighting = half3(input.fogFactorAndVertexLight.yzw);
+#endif
+
+#ifdef VARYINGS_NEED_SH
```

Comment:
You should exclude some case based on DECAL_DBUFFER or other define. Sample GI is not requried for decal buffer

Reply:
nevermind got confuse again :)